### DISCUSSION TITLE
Welcome to UoS_Beamforming Discussions!

### DISCUSSION THREAD

**Discussion by JoseAmador95** [JoseAmador95/UoS_Beamforming#1] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by johanflorez98** [comment]
Hello Jose. Thank you very much for sharing your work and giving us the opportunity to use it in different projects. My name is Johan Florez and I am an electronic engineer at the University of Nariño - Colombia. I am currently working on a research project where we are using the bladeRF radios used by you. I was looking for information that could help me to configure the transmission and reception of the 2 radio channels and luckily I found the information you shared. At the moment I have done some tests to validate the activation of both TX and RX channels, but I have a problem. When I activate both RX channels after a simulation time I get an error and the simulation stops and I have not been able to solve it. I would be infinitely grateful if with your experience you could help me to solve what happened or you could guide me on the possible causes or solutions. Thank you for taking the time to read me. I will be very attentive to your answer.

The error presented is: 

MATLAB System block 'bladeRF_TX_MIMO/MATLAB System' error occurred when invoking 'stepImpl' method of 'bladeRF_MIMO_Simulink'. The error was thrown from '
'C:\Users\ROBOTICA\Documents\bladeRF\matlab\bladeRF_MIMO.m' at line 116
'C:\Users\ROBOTICA\Documents\bladeRF\matlab\bladeRF_MIMO.m' at line 629
'C:\Users\ROBOTICA\Documents\bladeRF\matlab\bladeRF_MIMO_Simulink.m' at line 436'. 
* libbladeRF error (-6) in bladerf_sync_rx(): Operation timed out

**Reply by JoseAmador95** [reply]
Hi Johan! It has been a while since I worked with the bladeRF and I don't have access to the hardware at the moment. From helping another student, I realized that the release 2021.02 was not working with my Simulink models; I'm assuming you are using the newest 2021.10. You may try downgrading to 2019.07 or build a custom version including Warren's Pull Request (check the issue in this repository mentioning the PR).

Other thing I'd check is to set the runtime of **all the blocks** in your model to Generate code. Also, I found 1msps to be the sweet spot for my implementation (MATLAB/Simulink does add a considerable overhead). 

I am very willing to help. I see you are a Spanish speaker, so we can have a quick call.

**Reply by johanflorez98** [reply]
Muchas gracias Jose por tu respuesta. Revisaré las recomendaciones que me das y me estaré comunicando contigo según los resultados que obtenga. Muchas gracias nuevamente por tu disposición.

**Comment by mikedon1** [comment]
Hi Jose,
Thanks for publishing your code.  I'm thinking about using BladeRF for an angle of arrival application, and realize that you need to know the phase offset between the channels, as you mention on pg. 16 of your report.  You list two ways of doing this, either distributing a LO to every SDR, or distributing a low-freq. trigger signal to keep every LO aligned.  Could you give me more detail about how you did this?  I didn't find more information in your report, and it looks like my version of MATLAB can't open your simulink files.
Thanks, -Mike

**Reply by JoseAmador95** [reply]
Hi Mike!

For my setup in particular, the BladeRF provides coherent TX/RX channels out-of-the-box. So I really did no have to do anything in particular. The Simulink model in my repo should work as a DOA calculator as is with a single BladeRF.

Just to elaborate on your question, some boards like the BladeRf and LimeSDR do have clock input/output. These are basically low-frequency signals that are used to lock a PLL as I described in my report. Hence, you would be able to join two BladeRf boards together and create a 4x4 MIMO setup.

Other solutions, like [KrakenSDR](https://www.krakenrf.com/about-krakensdr) or the [homemade MIMO system based on Pluto SDR](https://www.youtube.com/@jonkraft), distribute the clock signal to all the RF chains.

Let me know if you have any other question!

Jose

**Reply by mikedon1** [reply]
OK, thanks for your reply, I thought the rx channels used a seperate LO even on the same chip, but looking into it more it looks like its what you've described.
